functions/dialect.go

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package functions
+
+import (
+	"fmt"
+	"github.com/goccy/go-yaml"
+	substraitgo "github.com/substrait-io/substrait-go"
+	"github.com/substrait-io/substrait-go/extensions"
+	"io"
+	"strings"
+)
+
+var loadedDialects = make(map[string]Dialect)
+
+func LoadDialect(name string, r io.Reader) error {
+	if _, ok := loadedDialects[name]; ok {
+		return fmt.Errorf("%w: dialect '%s' already loaded", substraitgo.ErrKeyExists, name)
+	}
+
+	dialect, err := newDialect(name, r)
+	if err != nil {
+		return err
+	}
+	loadedDialects[name] = dialect
+	return nil
+}
+
+func GetDialect(name string) (Dialect, error) {
+	if d, ok := loadedDialects[name]; ok {
+		return d, nil
+	}
+	return nil, fmt.Errorf("%w: dialect '%s' not found", substraitgo.ErrNotFound, name)
+}
+
+func GetSupportedDialects() []string {
+	var ret []string
+	for k := range loadedDialects {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+type dialectFunctionInfo struct {
+	ID        extensions.ID
+	Name      string
+	LocalName string
+	Options   map[string]extensions.Option
+	Notation  FunctionNotation
+}
+
+type dialectImpl struct {
+	name string
+	file dialectFile
+
+	toLocalTypeMap map[string]dialectTypeInfo // substrait short type name to dialectTypeInfo
+
+	localScalarFunctions    map[extensions.ID]*dialectFunctionInfo
+	localAggregateFunctions map[extensions.ID]*dialectFunctionInfo
+	localWindowFunctions    map[extensions.ID]*dialectFunctionInfo
+}
+
+func (d *dialectImpl) Name() string {
+	return d.name
+}
+
+func (d *dialectImpl) LocalizeFunctionRegistry(registry FunctionRegistry) (LocalFunctionRegistry, error) {
+	scalarFunctions := makeLocalFunctionVariantMap(d.localScalarFunctions, registry.GetScalarFunctions, newLocalScalarFunctionVariant)
+	aggregateFunctions := makeLocalFunctionVariantMap(d.localAggregateFunctions, registry.GetAggregateFunctions, newLocalAggregateFunctionVariant)
+	windowFunctions := makeLocalFunctionVariantMap(d.localWindowFunctions, registry.GetWindowFunctions, newLocalWindowFunctionVariant)
+
+	return newLocalFunctionRegistry(d, scalarFunctions, aggregateFunctions, windowFunctions), nil
+}
+
+type withID interface {
+	ID() extensions.ID
+}
+
+func makeLocalFunctionVariantMap[T withID, V any](dialectFunctionInfos map[extensions.ID]*dialectFunctionInfo, getFunctionVariants func(string) []T, createLocalVariant func(T, *dialectFunctionInfo) *V) map[string][]*V {
+	localFunctionVariants := make(map[string][]*V)
+	for _, dfi := range dialectFunctionInfos {
+		if _, ok := localFunctionVariants[dfi.Name]; ok {
+			// skip if function name is already added
+			continue
+		}
+		localVariantArray := make([]*V, 0)
+		for _, f := range getFunctionVariants(dfi.Name) {
+			if dfi, ok := dialectFunctionInfos[f.ID()]; ok {
+				// add functionVariant to local registry only if it is present in the dialect
+				localVariantArray = append(localVariantArray, createLocalVariant(f, dfi))
+			}
```

Comment:
We need to generate a collection of missing functions and produce an error of the missing functions. Right now it looks like we just ignore this case.